### ISSUE TITLE
Rule: Return or catch promises

### ISSUE THREAD

**Issue by andersekdahl** [eslint/eslint#1092]
Hi!
I'm making this rule for our team, but I thought it would probably make sense for other people as well.

It tries to detect if you have a promise, but you neither return it nor add an error handler ot it.

So this is good:

``` javascript
return promise.then(function() { ... });

promise.then(function() { ... }, function() { ... });

promise.then(function() { ... }).catch(function() { ... });
```

And this would raise an error:

``` javascript
promise.then(function() { ... });
```

Due to the dynamic nature of Javascript, this is quite tricky to statically analyze, since you can use all kinds of voodoo expressions. So the goal of this rule is not to let things like this through:

``` javascript
var m = 'cat';
promise.then(function() { ... })[m + 'ch'](function() { ... });
```

But it is the goal to detect as many valid cases as possible to not warn about valid code.

This rule should be turned off by default.

**Comment by xjamundx** [comment]
I'm :+1: something like this. @andersekdahl did you make any progress?

Basic thought is if you `then()` something you should `catch()` as well

**Comment by mik01aj** [comment]
[The `eslint-plugin-should-promised` plugin](https://github.com/dbrockman/eslint-plugin-should-promised) does something similar, but only for tests using `should`. Would be nice to have something like this for a generic use case. @andersekdahl, how's it going?

**Comment by AimeeKnight** [comment]
@andersekdahl I'm also curious if you've made progress on this? :+1:

**Comment by mugli** [comment]
@andersekdahl Any update on this? :)

**Comment by xjamundx** [comment]
Let me take a look at this today!

**Comment by xjamundx** [comment]
I'll make a plugin for this rule and then we can propose it back to core if it's useful. Will try to have it by mid-day tomorrow.

**Comment by xjamundx** [comment]
https://github.com/xjamundx/eslint-plugin-promise/

**Comment by mik01aj** [comment]
A bunch of comments from me:

* I'd like to have a rule that would work as proposed in the 1st post here - so that each promise either has errors handled with `catch` or is returned to another function.
* In these 3 examples, no two are equivalent:

   ```
   myPromise.then(doSomething, catchErrors); // catchErrors handles errors from myPromise
   myPromise.then(doSomething).catch(catchErrors);  // catchErrors handles errors from myPromise and doSomething
   myPromise.catch(catchErrors).then(doSomething);  // catchErrors handles errors from myPromise, and doSomething will operate on the result from catchErrors
   ```
   This kinda defeats the purpose of `always-catch` rule (at least in its current implementation).
